include/swift/AST/Types.h

```diff
@@ -4757,6 +4759,64 @@ class PrimaryArchetypeType final : public ArchetypeType,
 BEGIN_CAN_TYPE_WRAPPER(PrimaryArchetypeType, ArchetypeType)
 END_CAN_TYPE_WRAPPER(PrimaryArchetypeType, ArchetypeType)
 
+/// An archetype that represents an opaque type.
+class OpaqueTypeArchetypeType final : public ArchetypeType,
+    public llvm::FoldingSetNode,
+    private ArchetypeTrailingObjects<OpaqueTypeArchetypeType>
+{
+  friend TrailingObjects;
+  friend ArchetypeType;
+  friend GenericSignatureBuilder;
+
+  /// The declaration that defines the opaque type.
+  OpaqueTypeDecl *OpaqueDecl;
+  /// The substitutions into the interface signature of the opaque type.
+  SubstitutionMap Substitutions;
+
+  /// The generic signature builder used to build out this archetype, and its
+  /// nested archetypes.
+  
+public:
+  /// Get an opaque archetype representing the underlying type of the given
+  /// opaque type decl.
+  static OpaqueTypeArchetypeType *
+  get(OpaqueTypeDecl *Decl,
+      SubstitutionMap Substitutions);
+  
+  OpaqueTypeDecl *getOpaqueDecl() const {
+    return OpaqueDecl;
+  }
+  SubstitutionMap getSubstitutions() const {
+    return Substitutions;
+  }
+  
+  /// Get the underlying type of the opaque type, if it's known.
+  Type getUnderlyingType() const;
+  
+  static bool classof(const TypeBase *T) {
+    return T->getKind() == TypeKind::OpaqueTypeArchetype;
+  }
+  
+  static void Profile(llvm::FoldingSetNodeID &ID,
+                      OpaqueTypeDecl *OpaqueDecl,
+                      SubstitutionMap Substitutions);
+  
+  void Profile(llvm::FoldingSetNodeID &ID) {
+    Profile(ID, getOpaqueDecl(), getSubstitutions());
+  };
+  
+private:
+  OpaqueTypeArchetypeType(OpaqueTypeDecl *OpaqueDecl,
+                          SubstitutionMap Substitutions,
+                          RecursiveTypeProperties Props,
+                          GenericSignatureBuilder &&Builder,
```

Comment:
I'd prefer that we pass the `GenericSignature` itself (whose queries will eventually go through the GSB)` rather than the `GenericSignatureBuilder`. We want the GSB instances to be mostly invisible to the rest of the AST.